### PR TITLE
8265237: String.join and StringJoiner can be improved further

### PR DESCRIPTION
**Pull request by plevart**
While JDK-8148937 improved StringJoiner class by replacing internal use of getChars that copies out characters from String elements into a char[] array with StringBuilder which is somehow more optimal, the improvement was marginal in speed (0% ... 10%) and mainly for smaller strings, while GC was reduced by about 50% in average per operation.
Initial attempt to tackle that issue was more involved, but was later discarded because it was apparently using too much internal String details in code that lives outside String and outside java.lang package.
But there is another way to package such "intimate" code - we can put it into String itself and just call it from StringJoiner.
This PR is an attempt at doing just that. It introduces new package-private method in `java.lang.String` which is then used from both pubic static `String.join` methods as well as from `java.util.StringJoiner` (via SharedSecrets). The improvements can be seen by running the following JMH benchmark:

https://gist.github.com/plevart/86ac7fc6d4541dbc08256cde544019ce

The comparative results are here:

https://jmh.morethan.io/?gist=7eb421cf7982456a2962269137f71c15

The jmh-result.json files are here:

https://gist.github.com/plevart/7eb421cf7982456a2962269137f71c15

Improvement in speed ranges from 8% (for small strings) to 200% (for long strings), while creation of garbage has been further reduced to an almost garbage-free operation.

So WDYT?

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8265237](https://bugs.openjdk.java.net/browse/JDK-8265237): String.join and StringJoiner can be improved further

### Reviewers
 * [Roger Riggs](https://openjdk.java.net/census#rriggs) (@RogerRiggs - **Reviewer**) ⚠️ Review applies to a2d5e819fd72411909a300b775928ac837641361
 * [Claes Redestad](https://openjdk.java.net/census#redestad) (@cl4es - **Reviewer**) ⚠️ Review applies to a2d5e819fd72411909a300b775928ac837641361

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3501/head:pull/3501` \
`$ git checkout pull/3501`

Update a local copy of the PR: \
`$ git checkout pull/3501` \
`$ git pull https://git.openjdk.java.net/jdk pull/3501/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3501`

View PR using the GUI difftool: \
`$ git pr show -t 3501`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3501.diff">https://git.openjdk.java.net/jdk/pull/3501.diff</a>

</details>

### CODE REVIEW

**Comment by plevart** [comment]
Some background: This change was motivated by Sergey Tsypanov's attempt to replace usage of StringBuilder with string concatenation in JDK-8265075 only to find out that string concatenation in java.base module is compiled down to inline usage of StringBuilder, so no improvement was possible.
StringJoiner API and String.join static utility methods lend itself to a better implementation, but in last incarnation they are implemented with StringBuilder internally: String.join -> StringJoiner -> StringBuilder. A lot of JDK internal usages of StringBuilder were already replaced with StringJoiner (JDK-8054714) but under the hood the StringBuilder is still used. There were also lots of incremental attempts to improve StringJoiner. I think this one is the last one for some time to come... :-)

File: src/java.base/share/classes/java/lang/String.java [hunk 1]
```diff
@@ -3218,14 +3218,52 @@ public String replace(CharSequence target, CharSequence replacement) {
...
+        var delim = delimiter.toString();
+        var elems = new String[elements.length];
+        for (int i = 0; i < elements.length; i++) {
+            elems[i] = String.valueOf(elements[i]);
+        }
+        return join("", "", delim, elems, elems.length);
+    }
+
+    /**
+     * Designated join routine.
```
Comment (efge) [inline src/java.base/share/classes/java/lang/String.java:3230]: Did you mean "dedicated"?

**Comment by cl4es** [comment]
I think this seems reasonable. The String encoding details doesn't leak outside of java.lang, and numbers do look convincing.

There's a StringJoinerBenchmark micro added by JDK-8148937 which could perhaps be expanded with the scenarios you've experimented with here?

File: src/java.base/share/classes/java/lang/String.java [hunk 2]
```diff
@@ -3218,14 +3218,52 @@ public String replace(CharSequence target, CharSequence replacement) {
...
+     * @param prefix the non-null prefix
+     * @param suffix the non-null suffix
+     * @param delimiter the non-null delimiter
+     * @param elements the non-null array of non-null elements
+     * @param size the number of elements in the array (<= elements.length)
+     * @return the joined string
+     */
+    static String join(String prefix, String suffix, String delimiter, String[] elements, int size) {
+        int icoder = prefix.coder() | suffix.coder() | delimiter.coder();
+        long llen = (long) prefix.length() + suffix.length() + (long) Math.max(0, size - 1) * delimiter.length();
```
Comment (stsypanov) [inline src/java.base/share/classes/java/lang/String.java:3241]: If it's later cast to `int` why not to use `int` immediately here?

File: src/java.base/share/classes/java/lang/String.java (hunk 1 above)
Comment (plevart) [reply src/java.base/share/classes/java/lang/String.java:3230]: No, I meant designated. It is the routine that all other public API entry points call at the end to do the job. Would some other word more accurately describe that? I definitely didn't mean "dedicated".

Comment (efge) [reply src/java.base/share/classes/java/lang/String.java:3230]: Oh then sorry, I thought it was a typo of some sort. I'd have said something like "Centralized join logic". But whatever works for you.

**Comment by plevart** [comment]
> There's a StringJoinerBenchmark micro added by JDK-8148937 which could perhaps be expanded with the scenarios you've experimented with here?

I modified that micro benchmark and added a method to also measure String.join static method along with StringJoiner for same parameters and extended the range of parameters to cover more diversity. The results are here:

https://jmh.morethan.io/?gist=c38cc13d63774ec505cc8d394c00d502

It is apparent that there is a huge speedup when strings are bigger. But even smaller strings get a substantial speedup. There's also substantial reduction of garbage per operation. Previously the garbage amounted to the internal array of String elements and the StringBuffer with its internal byte[] array of characters. Now only the array of elements is the garbage.

File: src/java.base/share/classes/java/lang/String.java [hunk 3]
```diff
@@ -3218,14 +3218,52 @@ public String replace(CharSequence target, CharSequence replacement) {
...
+        int len = (int) llen;
+        if (llen != len) {
+            throw new OutOfMemoryError("Requested string length exceeds VM limit");
         }
-        return joiner.toString();
+
+        byte[] value = StringConcatHelper.newArray(((long) icoder << 32) | llen);
+        int off = 0;
+        prefix.getBytes(value, off, coder); off += prefix.length();
+        for (int i = 0; i < size; i++) {
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/lang/String.java:3256]: Can you save a branch inside the loop by handling element 0 outside the loop and
then do the loop for the rest?

File: src/java.base/share/classes/java/lang/String.java [hunk 4]
```diff
@@ -3218,14 +3218,52 @@ public String replace(CharSequence target, CharSequence replacement) {
...
+        }
+        byte coder = (byte) icoder;
+        int len = (int) llen;
+        if (llen != len) {
+            throw new OutOfMemoryError("Requested string length exceeds VM limit");
         }
-        return joiner.toString();
+
+        byte[] value = StringConcatHelper.newArray(((long) icoder << 32) | llen);
+        int off = 0;
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/lang/String.java:3254]: StringConcatHelper.newArray() can double the length (based on the coder) and it is then truncated to 32 bits when passed to UNSAFE.allocatlUnitializedArray.
The test of length above only ensures llen can be truncated to 32 bits without loss of data.

**Review by RogerRiggs — COMMENTED** [review]
Look very good.

File: src/java.base/share/classes/java/lang/String.java (hunk 3 above)
Comment (plevart) [reply src/java.base/share/classes/java/lang/String.java:3256]: Thanks, I'll do that and then re-test to see if there's any improvement.

File: src/java.base/share/classes/java/lang/String.java (hunk 4 above)
Comment (plevart) [reply src/java.base/share/classes/java/lang/String.java:3254]: I thought about that, yes. And I think we have to do the check for the doubled length before calling the newArray. I checked the StringJoinerTest and it only deals with ascii strings unfortunately. Will have to add a test for that too...

**Comment by amaembo** [comment]
Great work, thanks! I tried to do something similar a couple of years ago but did not submit my patch. One thing that stopped me is that joining many one-character strings was slower with my patch, compared to the baseline. Something like this:

```java
// setup
String[] strings = new String[100];
Arrays.fill(strings, "a");
// benchmark
return String.join(",", strings);
```

Could you add this case to your benchmark, for completeness?

**Comment by plevart** [comment]
So here's another iteration where I have taken into account all the suggestions:
- avoid the branch in the append loop by appending the 1st element outside the loop so that delimiter can unconditionally be appended inside the loop for rest of the elements.
- fixed overflow checking logic + added a test for it - previously it was possible to provoke wrong exception: IllegalArgumentException: size is negative. I added another test instead of extending existing one because it needs a 1GiB string and together with 2GiB string in existing test it would not fit in 4GiB heap any more when running with unpatched version. I wanted the tests to pass with unpatched version too.
- extended the parameters of JMH benchmark to include 1-character strings + re-run the benchmarks

The results are not that different from previous version (that used a conditional in the loop) as JIT is probably smart enough to unroll the loop and divide it into at least two parts (the way I did it manually now). What brings the most improvement in this version is the annotation to force in-lining of the String.join package-private method. Without it, it would happen that for one combination of JMH parameters, the results were a little worse compared to unpatched baseline - not for much: about 7%, but with the annotation, all combinations of parameters bring consistent improvement in range from 12% to 102%:

https://jmh.morethan.io/?gist=babf00a655d21ff784ede887af53ec31

File: src/java.base/share/classes/java/lang/String.java (hunk 4 above)
Comment (plevart) [reply src/java.base/share/classes/java/lang/String.java:3259]: I do the checks before calling `StringConcatHelper.newArray()` now and pass a long value to it that already holds the number of bytes needed and where the upper 32  bits (coder) is always 0.

File: test/micro/org/openjdk/bench/java/util/StringJoinerBenchmark.java [hunk 5]
```diff
@@ -41,9 +43,17 @@
  */
 @BenchmarkMode(Mode.AverageTime)
 @OutputTimeUnit(TimeUnit.NANOSECONDS)
-@Fork(jvmArgsAppend = {"-Xms2g", "-Xmx2g"})
+@Warmup(iterations = 10, time = 500, timeUnit = TimeUnit.MILLISECONDS)
+@Measurement(iterations = 10, time = 500, timeUnit = TimeUnit.MILLISECONDS)
+@Fork(value = 1, jvmArgsAppend = {"-Xms1g", "-Xmx1g"})
```
Comment (cl4es) [inline test/micro/org/openjdk/bench/java/util/StringJoinerBenchmark.java:48]: In general I think it's prudent to default to at least 3 forks, to catch issues with run-to-run variance.

**Review by cl4es — APPROVED** [review]
Looks good!

File: test/micro/org/openjdk/bench/java/util/StringJoinerBenchmark.java (hunk 5 above)
Comment (plevart) [reply test/micro/org/openjdk/bench/java/util/StringJoinerBenchmark.java:48]: Thanks, Claes. Will change that to 3 forks before pushing.
Any more comments? Will wait for a day, then push it.

**Review by RogerRiggs — APPROVED** [review]
Looks good.

**Pull request closed without merging**